transports/quic/src/hole_punching.rs

```diff
@@ -18,22 +19,26 @@ pub(crate) async fn hole_puncher<P: Provider>(
     futures::pin_mut!(punch_holes_future);
     match futures::future::select(P::sleep(timeout_duration), punch_holes_future).await {
         Either::Left(_) => Error::HandshakeTimedOut,
-        Either::Right((hole_punch_err, _)) => hole_punch_err,
+        Either::Right((Err(hole_punch_err), _)) => hole_punch_err,
+        Either::Right((Ok(never), _)) => match never {},
     }
 }
 
-async fn punch_holes<P: Provider>(socket: UdpSocket, remote_addr: SocketAddr) -> Error {
+async fn punch_holes<P: Provider>(
+    socket: UdpSocket,
+    remote_addr: SocketAddr,
+) -> Result<Infallible, Error> {
     loop {
-        let sleep_duration = Duration::from_millis(rand::thread_rng().gen_range(10..=200));
-        P::sleep(sleep_duration).await;
```

Comment:
@mxinden Sleeping here caused the tests to be quite flaky and probably impacted our real-world hole-punch success rate a lot as we effectively mitigated the `SYNC` part of DCUtR ..